packages/appium/lib/config-file.js

```diff
@@ -0,0 +1,293 @@
+      if (err.code === 'ENOENT') {
+        err.message = `Config file not found at user-provided path: ${filepath}`;
+      } else if (err instanceof SyntaxError) {
+        err.message = `Config file at user-provided path ${filepath} is invalid:\n${err.message}`;
+      }
+      throw err;
+    }
+  }
+  log.debug('No config file specified; searching...');
+  const result = await lc.search();
+  if (!result || !result.filepath) {
+    log.debug('Could not find a config file');
+  }
+  return result;
+}
+
+/**
+ * Inits Ajv, compiles the schema, and returns the validation function.
+ * Whenever this function is called, the validation function's `errors` property will be reset.
+ * If there are errors during validation, it will be set again.
+ */
+const validator = _.once(
+  /**
+   * @returns {import('ajv').ValidateFunction<import('./appium-config').AppiumConfigurationSchema>}
+   */
+  () => {
+    const ajv = addFormats(
+      new Ajv({
+        // without this not much validation actually happens
+        allErrors: true,
+        // enables use to use `"type": ["foo", "bar"]` in schema
+        allowUnionTypes: true,
+        // enables us to use custom properties (e.g., `appiumDest`)
+        strict: false,
+      }),
+    );
+    return ajv.compile(schema);
+  },
+);
+
+/**
+ * Validates a raw config object against the config schema.
+ * Initializes the Ajv singleton, if not already done.
+ * @param {object} [config] - Configuration object to validate. Could be anything.
+ * @returns {import('ajv').ErrorObject[]} Array of error objects from Ajv, if any.
+ */
+function validateConfigFile (config = {}) {
+  const validate = validator();
+  // validate.errors will be non-empty if `validate()` returns `false`.
+  // ...yes, that is a weird API.
+  return validate(config) ? [] : [...validate.errors];
+}
+
+/**
+ * Given an optional path, read a config file.
+ * @param {string} [filepath] - Path to config file, if we have one
+ * @param {ReadConfigFileOptions} [opts] - Options
+ * @returns {Promise<ReadConfigFileResult>} Contains config and filepath, if found, and any errors
+ */
+export async function readConfigFile (filepath, opts = {}) {
+  const result = (await findConfigFile(filepath)) ?? {};
+  if (result.config && !result.isEmpty) {
+    log.debug(`Config file found at ${result.filepath}`);
+    const {normalizeKeys = true, pretty = true} = opts;
+    /** @type {ReadConfigFileResult} */
+    let retval;
+    const errors = validateConfigFile(result.config);
+    if (errors.length) {
+      /** @type {string} */
+      let json;
+      try {
+        json = await fs.readFile(result.filepath, 'utf8');
+      } catch {}
+      const reason = betterAjvErrors(schema, result.config, errors, {
+        json,
+        format: pretty ? 'cli' : 'js',
+      });
+      retval = reason ? {...result, errors, reason} : {...result, errors};
+    } else {
+      retval = {...result, errors};
+    }
+    if (normalizeKeys) {
+      // XXX blah
```

Comment:
?

Reply:
I think I was unhappy about the magic array, as you noted...